### PR TITLE
Hunyang/dispatcher

### PR DESCRIPTION
**Pull request by hcyang**

### CODE REVIEW

File: packages/dispatcher/README.md [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+        `,confusionMatrixCrossValidation.getMacroAverageMetrics()=` +
+        `${confusionMatrixCrossValidation.getMacroAverageMetrics()}` +
+        `,confusionMatrixCrossValidation.getWeightedMacroAverageMetrics()=` +
+        `${confusionMatrixCrossValidation.getWeightedMacroAverageMetrics()}`);
+    return confusionMatrixCrossValidation;
+    // -----------------------------------------------------------------------
+}
+```
+
+<!-- ==== NOTE ==== PLACE-HOLDER-BEGIN ====
```
Comment (munozemilio) [inline packages/dispatcher/README.md]: Please delete the command related documentation from README

File: packages/dispatcher/package.json [hunk 2]
```diff
@@ -0,0 +1,62 @@
+{
+  "name": "@microsoft/bf-dispatcher",
+  "description": "Dispatcher contains a Softmax learner initially used for auto-active-learning down-sampling and a ML confusion-matrix evaluator on intent classification models.",
+  "version": "1.0.0",
+  "author": "Microsoft",
+  "bugs": "https://github.com/microsoft/botframework-cli/issues",
+  "dependencies": {
+    "@microsoft/bf-lu": "^4.6.0",
```
Comment (munozemilio) [inline packages/dispatcher/package.json]: please set this as: "@microsoft/bf-lu": "1.0.0"

File: packages/dispatcher/README.md (hunk 1 above)
Comment (hcyang) [reply packages/dispatcher/README.md]: Those lines were boiler template code auto generated. Had them be deleted, instead of commented, I would not know how to recover them when it's ready to create a new command.

File: packages/dispatcher/package.json (hunk 2 above)
Comment (hcyang) [reply packages/dispatcher/package.json]: missed one, will revise it.

File: packages/dispatcher/src/commands/hello.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import {Command, flags} from "@oclif/command";
```
Comment (munozemilio) [inline packages/dispatcher/src/commands/hello.ts]: Please delete this file since it wont be used

File: packages/dispatcher/data/Columnar/Email.tsv.metrics.json [hunk 4]
```diff
@@ -0,0 +1,328 @@
+{
```
Comment (munozemilio) [inline packages/dispatcher/data/Columnar/Email.tsv.metrics.json]: Is this test data? if so can we rename the container folder 'data' to something else like 'fixtures' or 'testdata' and probably move it under the test folder?? This is bc there is already a data folder in src.

File: packages/dispatcher/README.md (hunk 1 above)
Comment (munozemilio) [reply packages/dispatcher/README.md]: You just need to leave:

When the commands are implemented the documentation will be autogenerated from there

File: packages/dispatcher/package.json [hunk 5]
```diff
@@ -0,0 +1,62 @@
+{
```
Comment (munozemilio) [inline packages/dispatcher/package.json]: main is missing. What is the entry point of the library?

File: packages/dispatcher/README.md (hunk 1 above)
Comment (hcyang) [reply packages/dispatcher/README.md]: Will backup those commented commands and remove them from the PR.

File: packages/dispatcher/data/Columnar/Email.tsv.metrics.json (hunk 4 above)
Comment (hcyang) [reply packages/dispatcher/data/Columnar/Email.tsv.metrics.json]: The dataset files in the 'data' folder are used for examples, unit tests, and demo. 'testdata' has special meaning in ML, so not a good fit. 'fixtures' is not clear what it is. Trying to think if there is a better name. Is it a problem having a sub 'data' folder specific for dispatcher?

File: packages/dispatcher/src/commands/hello.ts (hunk 3 above)
Comment (hcyang) [reply packages/dispatcher/src/commands/hello.ts]: Archived and removed.

File: packages/dispatcher/package.json (hunk 5 above)
Comment (hcyang) [reply packages/dispatcher/package.json]: It's a library, so instead of an entry point, there are plenty of little apps to demonstrate how the library can be used.

File: packages/dispatcher/data/Columnar/Email.tsv.metrics.json (hunk 4 above)
Comment (munozemilio) [reply packages/dispatcher/data/Columnar/Email.tsv.metrics.json]: This will do the code easier to maintain. Also since this is used for unit test and demo it makes sense to move it under test folder

File: packages/dispatcher/package.json (hunk 5 above)
Comment (munozemilio) [reply packages/dispatcher/package.json]: The main property of a package.json is a direction to the entry point to the module that the package.json is describing. This is what the users of the library will be consuming.

Comment (hcyang) [reply packages/dispatcher/package.json]: No, the users should read README.md. package.json is for maintainers.

File: packages/dispatcher/data/Columnar/Email.tsv.metrics.json (hunk 4 above)
Comment (hcyang) [reply packages/dispatcher/data/Columnar/Email.tsv.metrics.json]: This is really a minor issue. How about I will create a 'resources' folder and put the data files under?

File: packages/dispatcher/package.json [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  "name": "@microsoft/bf-dispatcher",
+  "description": "Dispatcher contains a Softmax learner initially used for auto-active-learning down-sampling and a ML confusion-matrix evaluator on intent classification models.",
+  "version": "1.0.0",
+  "author": "Microsoft",
+  "bugs": "https://github.com/microsoft/botframework-cli/issues",
+  "dependencies": {
+    "@microsoft/bf-lu": "^1.0.0",
```
Comment (munozemilio) [inline packages/dispatcher/package.json]: Please set it to be exactly like this:  "@microsoft/bf-lu": "1.0.0"

File: packages/dispatcher/package.json [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    "@oclif/command": "^1.5.19",
+    "@oclif/config": "^1.13.3",
+    "tslib": "^1.10.0"
+  },
+  "devDependencies": {
+    "@microsoft/bf-lu": "^1.0.0",
```
Comment (munozemilio) [inline packages/dispatcher/package.json]: no needed in here

File: packages/dispatcher/data/Columnar/Email.tsv.metrics.json (hunk 4 above)
Comment (hcyang) [reply packages/dispatcher/data/Columnar/Email.tsv.metrics.json]: created a new folder "resources"

File: packages/dispatcher/package.json (hunk 6 above)
Comment (hcyang) [reply packages/dispatcher/package.json]: revised.

File: packages/dispatcher/package.json (hunk 7 above)
Comment (hcyang) [reply packages/dispatcher/package.json]: removed

File: packages/dispatcher/package.json (hunk 5 above)
Comment (hcyang) [reply packages/dispatcher/package.json]: populated index.ts

Comment (hcyang) [reply packages/dispatcher/package.json]: and point the main entry in package.json to its js version

**Comment by munozemilio** [comment]
@christopheranderson I just updated dispatcher with the new rush scripts, could you help me double check everything looks good please?

File: packages/dispatcher/README.md [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+        true): ConfusionMatrix {
+    // -----------------------------------------------------------------------
+    const columnarData: ColumnarData =
+        ColumnarData.createColumnarData(
+            columnarContent,
+            new NgramSubwordFeaturizer(),
+            labelColumnIndex,
+            textColumnIndex,
+            linesToSkip);
+    // -----------------------------------------------------------------------
```
Comment (munozemilio) [inline packages/dispatcher/README.md]: Is this intended to be commented in the README?

File: packages/dispatcher/src/data/AppColumnarData.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        },
+    );
+    parser.addArgument(
+        ["-d", "--debug"],
+        {
+            help: "enable printing debug information",
+            required: false,
+        },
+    );
+    // parser.addArgument(
```
Comment (munozemilio) [inline packages/dispatcher/src/data/AppColumnarData.ts]: Should we delete this?

File: packages/dispatcher/src/data/AppColumnarData.ts [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+import { NgramSubwordFeaturizer } from "../model/language_understanding/featurizer/NgramSubwordFeaturizer";
+
+import { Utility } from "../utility/Utility";
+
+export function exampleFunctionData(): ColumnarData {
+    // -----------------------------------------------------------------------
+    Utility.debuggingLog(`process.argv=${process.argv}`);
+    // -----------------------------------------------------------------------
+    const parser = new ArgumentParser({
```
Comment (munozemilio) [inline packages/dispatcher/src/data/AppColumnarData.ts]: Is this intended to be a command?

File: packages/dispatcher/src/data/AppEntityAnnotatedCorpusData.ts [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        },
+    );
+    parser.addArgument(
+        ["-d", "--debug"],
+        {
+            help: "enable printing debug information",
+            required: false,
+        },
+    );
+    // parser.addArgument(
```
Comment (munozemilio) [inline packages/dispatcher/src/data/AppEntityAnnotatedCorpusData.ts]: Should this be removed?

File: packages/dispatcher/src/data/AppEntityAnnotatedCorpusData.ts [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+import { ArgumentParser } from "argparse";
+
+import { EntityAnnotatedCorpusData } from "./EntityAnnotatedCorpusData";
+
+import { NgramSubwordFeaturizer } from "../model/language_understanding/featurizer/NgramSubwordFeaturizer";
+
+import { Utility } from "../utility/Utility";
+
+export function exampleFunctionData(): EntityAnnotatedCorpusData {
+    // -----------------------------------------------------------------------
```
Comment (munozemilio) [inline packages/dispatcher/src/data/AppEntityAnnotatedCorpusData.ts]: This has some duplicated code from AppColumnarData.ts. Is it possible to refactor to avoid code duplication?

File: packages/dispatcher/src/data/AppEntityAnnotatedCorpusData.ts [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    Utility.debuggingLog(`process.argv=${process.argv}`);
+    // -----------------------------------------------------------------------
+    const parser = new ArgumentParser({
+        addHelp: true,
+        description: "AppEntityAnnotatedCorpusData",
+        version: "0.0.1",
+    });
+    parser.addArgument(
```
Comment (munozemilio) [inline packages/dispatcher/src/data/AppEntityAnnotatedCorpusData.ts]: Is this a command? if it is, it should use the oclif way to implement the commands, if it is not, it would be nice to decouple the library functionality from the command one.

File: packages/dispatcher/src/data/EntityAnnotatedCorpusData.ts [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+
+    public getLinesToSkip(): number {
+        return this.linesToSkip;
+    }
+
+    public dumpLuObject(
+        filename: string,
+        replacer?: (this: any, key: string, value: any) => any,
+        space?: string | number): void {
+        // Utility.dumpFile(
```
Comment (munozemilio) [inline packages/dispatcher/src/data/EntityAnnotatedCorpusData.ts]: If there is no TODO pending we should delete this code

File: packages/dispatcher/src/data/EntityAnnotatedCorpusData.ts [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+        if (toResetFeaturizerLabelFeatureMaps) {
+            entityAnnotatedCorpusData.resetFeaturizerLabelFeatureMaps();
+        }
+        // -------------------------------------------------------------------
+        entityAnnotatedCorpusData.featurizeIntentsUtterances();
+        // -------------------------------------------------------------------
+        return entityAnnotatedCorpusData;
+    }
+
+    // protected luObject: any = null;
```
Comment (munozemilio) [inline packages/dispatcher/src/data/EntityAnnotatedCorpusData.ts]: Still needed?

File: packages/dispatcher/README.md (hunk 8 above)
Comment (hcyang) [reply packages/dispatcher/README.md]: Status check code, will un-comment them

File: packages/dispatcher/src/data/AppColumnarData.ts (hunk 9 above)
Comment (hcyang) [reply packages/dispatcher/src/data/AppColumnarData.ts]: They are placeholder for future use. Add a comment.

File: packages/dispatcher/src/data/AppColumnarData.ts (hunk 10 above)
Comment (hcyang) [reply packages/dispatcher/src/data/AppColumnarData.ts]: They will become a command once the PMs have settled a plan for transitioning 'dispatcher' as a replacement for C# Dispatch. As for now,  'dispatcher' is designed as a library, and those are example functions that can be referenced by Composer.

File: packages/dispatcher/src/data/AppEntityAnnotatedCorpusData.ts (hunk 11 above)
Comment (hcyang) [reply packages/dispatcher/src/data/AppEntityAnnotatedCorpusData.ts]: They are placeholder code that will be activated in the future. Added a comment to indicate that.

File: packages/dispatcher/src/data/AppEntityAnnotatedCorpusData.ts (hunk 12 above)
Comment (hcyang) [reply packages/dispatcher/src/data/AppEntityAnnotatedCorpusData.ts]: The two files are examples and quite small. Would be over-engineering if only to refactor the common argument processing code.

File: packages/dispatcher/src/data/AppEntityAnnotatedCorpusData.ts (hunk 13 above)
Comment (hcyang) [reply packages/dispatcher/src/data/AppEntityAnnotatedCorpusData.ts]: 'dispatcher' currently is a library, thus there are plenty of example code for reference purpose. These example functions can be upgraded to OCLIF commands after (in the fuure) the PMs settle down a plan to transition 'dispatcher' to replacing C# Dispatch CLI.

File: packages/dispatcher/src/data/EntityAnnotatedCorpusData.ts (hunk 14 above)
Comment (hcyang) [reply packages/dispatcher/src/data/EntityAnnotatedCorpusData.ts]: Refactored code deleted, TODO comment added.

File: packages/dispatcher/src/data/EntityAnnotatedCorpusData.ts (hunk 15 above)
Comment (hcyang) [reply packages/dispatcher/src/data/EntityAnnotatedCorpusData.ts]: Will remove.

File: packages/dispatcher/.travis.yml [hunk 16]
```diff
@@ -0,0 +1,10 @@
+language: node_js
```
Comment (christopheranderson) [inline packages/dispatcher/.travis.yml]: delete this file

File: packages/dispatcher/package.json [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+    "chai": "^4.2.0",
+    "globby": "^10.0.1",
+    "mocha": "^5.2.0",
+    "nyc": "^14.1.1",
+    "readline-sync": "^1.4.10",
+    "rimraf": "^3.0.0",
+    "ts-md5": "^1.2.6",
+    "ts-node": "^8.4.1",
+    "tslint": "^5.20.1",
+    "typescript": "^3.7.2"
```
Comment (christopheranderson) [inline packages/dispatcher/package.json]: different versions of TypeScript from other libraries - @munozemilio, we have a couple of versions. I recommend we stick with 3.5.x unless we have a compelling reason to upgrade.

File: packages/dispatcher/src/data/AppColumnarData.ts [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+        labelColumnIndex,
+        textColumnIndex,
+        linesToSkips,
+        true);
+    // columnarData.dumpLuJsonStructure(outputFilename);
+    return columnarData;
+    // -----------------------------------------------------------------------
+}
+
+if (require.main === module) {
```
Comment (christopheranderson) [inline packages/dispatcher/src/data/AppColumnarData.ts]: why is this here?

File: packages/dispatcher/src/model/evaluation/cross_validation/AppCrossValidator.ts [hunk 19]
```diff
@@ -0,0 +1,543 @@
...
+                numberOfCrossValidationFolds,
+                learnerParameterEpochs,
+                learnerParameterMiniBatchSize,
+                learnerParameterL1Regularization,
+                learnerParameterL2Regularization,
+                learnerParameterLossEarlyStopRatio,
+                learnerParameterLearningRate,
+                learnerParameterToCalculateOverallLossAfterEpoch);
+    }
+    const confusionMatrixMetricStructure: { "confusionMatrix": ConfusionMatrix,
```
Comment (christopheranderson) [inline packages/dispatcher/src/model/evaluation/cross_validation/AppCrossValidator.ts]: this is a very complicated inline type. Can this be extracted?

File: packages/dispatcher/src/data/AppColumnarData.ts (hunk 18 above)
Comment (hcyang) [reply packages/dispatcher/src/data/AppColumnarData.ts]: It's an example function demonstrating ways to use ColumnData objects.

File: packages/dispatcher/.travis.yml (hunk 16 above)
Comment (hcyang) [reply packages/dispatcher/.travis.yml]: deleted.

File: packages/dispatcher/src/model/evaluation/cross_validation/AppCrossValidator.ts (hunk 19 above)
Comment (hcyang) [reply packages/dispatcher/src/model/evaluation/cross_validation/AppCrossValidator.ts]: Can be refactored in next check-in

**Comment by hcyang** [comment]
initial PR merge for the dispatcher package

File: packages/dispatcher/src/mathematics/sampler/BootstrapSamplerKeyMap.ts [hunk 20]
```diff
@@ -0,0 +1,111 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+// import { MathematicsHelper } from "../mathematics_helper/MathematicsHelper";
```
Comment (daveta) [inline packages/dispatcher/src/mathematics/sampler/BootstrapSamplerKeyMap.ts]: >import { MathematicsHelper [](start = 3, length = 26)

This seems to be included and commented out alot.. Just curious if this should be removed altogether?

File: packages/dispatcher/src/model/evaluation/abstract_base_evaluator/AbstractBaseEvaluator.ts [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IDictionaryStringIdGenericArrays } from "../../../data_structure/IDictionaryStringIdGenericArrays";
+import { IDictionaryStringIdGenericValue } from "../../../data_structure/IDictionaryStringIdGenericValue";
+
+import { Utility } from "../../../utility/Utility";
+
+// import { Utility } from "../../../utility/Utility";
```
Comment (daveta) [inline packages/dispatcher/src/model/evaluation/abstract_base_evaluator/AbstractBaseEvaluator.ts]: Hmm.

File: packages/dispatcher/src/mathematics/sampler/BootstrapSamplerKeyMap.ts (hunk 20 above)
Comment (hcyang) [reply packages/dispatcher/src/mathematics/sampler/BootstrapSamplerKeyMap.ts]: will remove them.

File: packages/dispatcher/src/model/evaluation/abstract_base_evaluator/AbstractBaseEvaluator.ts (hunk 21 above)
Comment (hcyang) [reply packages/dispatcher/src/model/evaluation/abstract_base_evaluator/AbstractBaseEvaluator.ts]: Will remove them
